File: backend/app/modules/tenants/entitlements.py
from dataclasses import asdict, dataclass

from fastapi import HTTPException

from app.core.config import settings


@dataclass(frozen=True)
class Entitlements:
    is_unlimited: bool
    message_quota: int | None
    tg_account_limit: int | None
    mapping_limit: int | None
    bot_limit: int | None
    member_limit: int | None

    def as_dict(self) -> dict:
        return asdict(self)


VIEWER_FEATURES = {"dashboard", "telegram", "dingtalk", "routes"}
UNLIMITED_PLANS = {"pro", "enterprise"}


def resolve_entitlements(user: dict, plan: str) -> Entitlements:
    role = user.get("role", "member")
    if role == "viewer" and not user.get("is_platform_admin", False):
        plan_limits = settings.PLAN_LIMITS.get(plan, settings.PLAN_LIMITS["free"])
        return Entitlements(False, plan_limits["message_quota"], 1, 1, 1, 0)

    unlimited = bool(user.get("is_platform_admin")) or plan in UNLIMITED_PLANS
    if unlimited:
        return Entitlements(True, None, None, None, None, None)

    limits = settings.PLAN_LIMITS.get(plan, settings.PLAN_LIMITS["free"])
    return Entitlements(False, **limits)


def require_feature(user: dict, feature: str) -> None:
    if user.get("is_platform_admin") or user.get("role") != "viewer":
        return
    if feature not in VIEWER_FEATURES:
        raise HTTPException(403, "观察者无权访问此功能")


def require_authorization_admin(user: dict) -> None:
    if not user.get("is_platform_admin"):
        raise HTTPException(403, "只有平台管理员可以变更授权")

File: frontend/src/lib/entitlements.ts
export interface EntitlementSummary {
  plan?: string
  role?: string
  is_platform_admin?: boolean
  is_unlimited?: boolean
  usage_percent?: number | null
}

export function limitLabel(value: number | null | undefined, isUnlimited = false): string {
  return isUnlimited ? '∞' : String(value ?? 0)
}

export function planLabel(summary: EntitlementSummary): string {
  if (!summary.is_unlimited) return summary.plan || 'free'
  if (summary.is_platform_admin || (!summary.role && summary.plan === 'free')) return '平台无限'
  if (summary.plan === 'pro') return '专业版无限'
  if (summary.plan === 'enterprise') return '企业版无限'
  return '管理员无限'
}

export function usageNote(summary: EntitlementSummary): string {
  return summary.is_unlimited ? '不限制' : `${summary.usage_percent ?? 0}% 已使用`
}

export function planPrice(plan: string): string {
  if (plan === 'pro') return '¥19,800'
  if (plan === 'free') return '免费'
  return '联系平台管理员'
}
